### PR TITLE
Add job constraints

### PR DESCRIPTION
**Pull request by wyegelwel**
This adds the ability to set constraints on individual jobs and provides a way to grow the constraints api in the future. It is inspired by marathon's constraint api: https://mesosphere.github.io/marathon/docs/constraints.html

### CODE REVIEW

**Comment by wyegelwel** [comment]
@shamsimam @sradack , any comments on this?

**Comment by shamsimam** [comment]
Having a look now

File: integration/tests/cook/test_basic.py [hunk 1]
```diff
@@ -306,6 +306,29 @@ def test_error_while_creating_job(self):
         resp = util.session.post('%s/rawscheduler' % self.cook_url,
                                  json={'jobs':[job1, job2]})
         self.assertEqual(resp.status_code, 500)
+    
+    def test_constraints(self):
+        # Want to ensure job is put on specific host, but no way to set host names in minimesos
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Can we add a test where we use the host name from `http://mesos-master:5050/state.json` user `slaves->*->hostname` ?

Then you can verify whether exactly that host was used. Maybe launch multiple jobs with different host constraints and verify each job only ran on the specified host?

File: scheduler/docs/scheduler-rest-api.asc [hunk 2]
```diff
@@ -113,12 +114,23 @@ The request body is list of job maps with each entry containing a map of the fol
...
 ----
 
+===== Constraints
+
+Constraints provide controls over where a job is placed on the cluster. 
+There are both job and group level constraints.
+This will discuss job constraints, see the docs/groups.md for more details on group constraints.
+
+Constraints are specified as a tuple of attribute, operator, pattern.
+Attribute can be any attribute set on a host in the cluster.
```
Comment (shamsimam) [inline scheduler/docs/scheduler-rest-api.asc]: Is there a means to query cook to find out available attributes on agents?

File: scheduler/src/cook/mesos/api.clj [hunk 3]
```diff
@@ -28,7 +28,7 @@
             [compojure.core :refer (ANY GET POST routes)]
             [cook.mesos.quota :as quota]
             [cook.mesos.reason :as reason]
-            [cook.mesos.schema :refer (host-placement-types straggler-handling-types)]
+            [cook.mesos.schema :refer (host-placement-types straggler-handling-types constraint-operators)]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Nitpick: lexicographic ordering :(

File: scheduler/src/cook/mesos/api.clj [hunk 4]
```diff
@@ -450,6 +461,16 @@
                              :label/key k
                              :label/value v}]))
                        labels)
+        constraints (mapcat (fn [[attribute operator pattern]]
+                              (let [constraint-var-id (d/tempid :db.part/user)]
+                                [[:db/add db-id :job/constraint constraint-var-id]
+                                 (merge {:db/id constraint-var-id
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Can we use `cond->` and `assoc` instead of `merge`?

File: scheduler/src/cook/mesos/api.clj [hunk 5]
```diff
@@ -609,6 +632,9 @@
                                uris)})
                  (when labels
                    {:labels (walk/stringify-keys labels)})
+                 (when constraints
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Nitpick: another good spot for `cond->` amd `assoc`

File: scheduler/src/cook/mesos/api.clj [hunk 6]
```diff
@@ -740,6 +766,13 @@
                   "success" "failed")
                 :job.state/running "running"
                 :job.state/waiting "waiting")
+        constraints (->> job
+                         :job/constraint
+                         (map util/remove-datomic-namespacing)
+                         (map (fn [{:keys [attribute operator pattern]}]
+                                (->> [attribute (clojure.string/upper-case (name (util/without-ns operator))) pattern]
+                                     (remove nil?)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Why can we have `nil` values in the constraint? Doesn't the schema protect against that?

File: scheduler/src/cook/mesos/constraints.clj [hunk 7]
```diff
@@ -116,7 +117,37 @@
...
+                                      pattern :constraint/pattern
+                                      operator :constraint/operator}]
+            (let [vm-attribute-value (get vm-attributes attribute)]
+              (condp = operator
+                :constraint.operator/equals (= pattern vm-attribute-value)
+                :else (do
+                        (log/error (str "Unknown operator " operator
+                                        " api.clj should have prevented this from happening."))
+                        true))))
+          constraint->passes? (pc/map-from-keys vm-passes-constraint? constraints)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Why are we building a map here? Couldn't we directly do: `(every? vm-passes-constraint? constraints)`?

File: scheduler/test/cook/test/mesos/constraints.clj [hunk 8]
```diff
@@ -33,6 +33,15 @@
   (is (= "attribute-equals-host-placement-group-constraint"
          (constraints/group-constraint-name (constraints/->attribute-equals-host-placement-group-constraint nil)))))
 
+(deftest test-user-defined-constraint
+  (let [constraints [{:constraint/attribute "is_spot"
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/constraints.clj]: We should also add test assertions for multiple constraints scenarios.

File: scheduler/src/cook/mesos/api.clj [hunk 9]
```diff
@@ -450,6 +461,16 @@
                              :label/key k
                              :label/value v}]))
                        labels)
+        constraints (mapcat (fn [[attribute operator pattern]]
+                              (let [constraint-var-id (d/tempid :db.part/user)]
+                                [[:db/add db-id :job/constraint constraint-var-id]
+                                 (merge {:db/id constraint-var-id
+                                         :constraint/attribute attribute
+                                         :constraint/operator (keyword "constraint.operator"
+                                                                       (clojure.string/lower-case operator))}
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Nitpick: can we require `clojure.string :as str`

File: scheduler/src/cook/mesos/api.clj [hunk 10]
```diff
@@ -740,6 +766,13 @@
                   "success" "failed")
                 :job.state/running "running"
                 :job.state/waiting "waiting")
+        constraints (->> job
+                         :job/constraint
+                         (map util/remove-datomic-namespacing)
+                         (map (fn [{:keys [attribute operator pattern]}]
+                                (->> [attribute (clojure.string/upper-case (name (util/without-ns operator))) pattern]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Nitpick: can we require `clojure.string :as str`

File: scheduler/src/cook/mesos/rebalancer.clj [hunk 11]
```diff
@@ -232,7 +232,8 @@
                              :normal (dru/sorted-task-scored-task-pairs user->dru-divisors user->sorted-running-task-ents)
                              :gpu (dru/sorted-task-cumulative-gpu-score-pairs user->dru-divisors user->sorted-running-task-ents))
          task->scored-task (into (pm/priority-map-keyfn (case category
-                                                          :normal (comp - :dru)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/rebalancer.clj]: Why did this change?

File: scheduler/src/cook/mesos/constraints.clj [hunk 12]
```diff
@@ -116,7 +117,37 @@
   (let [needs-gpus? (job-needs-gpus? job)]
     (->gpu-host-constraint job needs-gpus?)))
 
-(def job-constraint-constructors [build-novel-host-constraint build-gpu-host-constraint])
+(defrecord user-defined-constraint [constraints]
+  JobConstraint
+  (job-constraint-name [this] (get-class-name this))
+  (job-constraint-evaluate
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Do we also need to implement the overloaded variant?

File: scheduler/src/cook/mesos/api.clj (hunk 6 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: The pattern is optional in the schema. I can make it required for now since the only operator needs it. Would you prefer that?

File: scheduler/src/cook/mesos/rebalancer.clj (hunk 11 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/rebalancer.clj]: Good question. It makes the test deterministic.

File: scheduler/src/cook/mesos/api.clj (hunk 6 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/api.clj]: Do you have example constraints where the pattern may be optional. If so, then okay, else I would prefer it to be required.

Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: I was using the same set up as marathon's constraints, https://mesosphere.github.io/marathon/docs/constraints.html which have an optional pattern, but looking more closely they blend job and group constraints in a way that I'm not sure we want. I will make it required.

File: scheduler/src/cook/mesos/rebalancer.clj [hunk 13]
```diff
@@ -232,7 +232,8 @@
...
+                                                          :normal (juxt (comp - :dru)
+                                                                        (comp util/task-ent->user :task))
                                                           :gpu (fnil - 0)))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/rebalancer.clj]: Does the function for `:gpu` now need to return a vector?

Comment (wyegelwel) [inline scheduler/src/cook/mesos/rebalancer.clj]: It isn't necessary, gpu will just not have a tie break when drus are equal. I only added it for normal because there are some tests that would fail when you changed unrelated source.

File: scheduler/src/cook/mesos/constraints.clj (hunk 12 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/constraints.clj]: I think I may misunderstand the question. I *think* we do.

File: scheduler/src/cook/mesos/constraints.clj (hunk 7 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/constraints.clj]: Good point, I was going to say which constraints failed but decided against it because it causes a lot of memory allocs.

File: scheduler/src/cook/mesos/constraints.clj (hunk 12 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: There is an overloaded version: `[this target-vm-resources target-vm-attrs target-vm-tasks-assigned]` which we are not implementing here. I'm not sure where the overloaded version is implemented?

File: scheduler/src/cook/mesos/api.clj (hunk 5 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: Would rather do that in a separate pr, there are a lot of these conditions that would change

File: scheduler/docs/scheduler-rest-api.asc (hunk 2 above)
Comment (wyegelwel) [inline scheduler/docs/scheduler-rest-api.asc]: At the moment, no. That is a good thing to expose. Adding issue for it: #370

File: scheduler/src/cook/mesos/constraints.clj [hunk 14]
```diff
@@ -17,6 +17,7 @@
   (:require [clojure.tools.logging :as log]
             [cook.mesos.group :as group]
             [cook.mesos.util :as util]
+            [plumbing.core :as pc]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: This import is unused now.

File: scheduler/src/cook/mesos/constraints.clj (hunk 12 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Saw it in line 130, my bad.

**Comment by wyegelwel** [comment]
Ok, removed the import and used mesos state.json to get the list of hosts and submitted a job constrained to each host.

File: integration/tests/cook/test_basic.py [hunk 15]
```diff
@@ -308,24 +309,26 @@ def test_error_while_creating_job(self):
...
-        self.assertEqual(hostname, hostname_constrained)
-        self.assertEqual(constraints, job['constraints'])
+        host_to_job_uuid = {}
+        for hostname in hosts:
+            constraints = [["HOSTNAME", "EQUALS", hostname]]
+            job_uuid, resp = util.submit_job(self.cook_url, constraints=constraints)
+            self.assertEqual(resp.status_code, 201, resp.text)
+            host_to_job_uuid[hostname] = job_uuid
+
+        for hostname, job_uuid in host_to_job_uuid.items():
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: +1

**Comment by wyegelwel** [comment]
Any more feedback on this or can we merge?

**Comment by shamsimam** [comment]
I approved. I saw @sradack as another reviewer, so didn't merge.

File: travis/minimesosFile [hunk 16]
```diff
@@ -2,7 +2,7 @@ minimesos {
     clusterName = "Cook Simulation Cluster"
     loggingLevel = "INFO"
     mapAgentSandboxVolume = true
-    mapPortsToHost = false
+    mapPortsToHost = true
```
Comment (sradack) [inline travis/minimesosFile]: Why is this change being made on this ?

File: scheduler/src/cook/mesos/api.clj [hunk 17]
```diff
@@ -194,6 +194,15 @@
   {:name (s/constrained s/Str non-empty-max-128-characters-and-alphanum?)
    :version (s/constrained s/Str non-empty-max-128-characters-and-alphanum?)})
 
+(def Constraint
+  "Schema for user defined job host constraint"
+  [(s/one NonEmptyString "attribute")
+   (s/one (s/pred #(contains? (set (map (comp name util/without-ns) constraint-operators))
```
Comment (sradack) [inline scheduler/src/cook/mesos/api.clj]: Can `(comp name util/without-ns)` be simply `name`?

File: scheduler/src/cook/mesos/api.clj [hunk 18]
```diff
@@ -740,6 +766,12 @@
                   "success" "failed")
                 :job.state/running "running"
                 :job.state/waiting "waiting")
+        constraints (->> job
+                         :job/constraint
+                         (map util/remove-datomic-namespacing)
+                         (map (fn [{:keys [attribute operator pattern]}]
+                                (->> [attribute (str/upper-case (name (util/without-ns operator))) pattern]
```
Comment (sradack) [inline scheduler/src/cook/mesos/api.clj]: Can `(name (util/without-ns operator))` be simply `(name operator)`?

File: travis/minimesosFile (hunk 16 above)
Comment (sradack) [inline travis/minimesosFile]: Shams mentioned this is for the integration test.

Comment (wyegelwel) [inline travis/minimesosFile]: I wasn't sure how to find the IP for zookeeper otherwise. Also, it mirrors the dev set up so it is less likely to cause problems where in dev tests pass but in integration they don't

**Comment by sradack** [comment]
Additionally there are some comments from @shamsimam that have gone without a response.

File: scheduler/src/cook/mesos/api.clj (hunk 17 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: Oh wow, yes. I didn't know name would strip the namespace as well. Will change now

File: scheduler/test/cook/test/mesos/constraints.clj (hunk 8 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/constraints.clj]: Done

File: scheduler/src/cook/mesos/api.clj (hunk 4 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: Since pattern is no longer optional, just going to add :pattern into the map

File: scheduler/src/cook/mesos/api.clj (hunk 17 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: Looking at the implementation of `without-ns` I clearly knew that at some point...

**Comment by wyegelwel** [comment]
Ok, responded to comments. Thanks for pointing out I missed one of Shams!

**Comment by wyegelwel** [comment]
Let me know if there is more feedback. Otherwise, may we merge?
